File: src/lib/copilot.js
/**
 * GitHub Copilot OAuth device flow + JWT management.
 *
 * Auth is two-layer:
 *  1. OAuth token  — long-lived, obtained via device flow, stored via ai_store_key
 *  2. Copilot JWT  — short-lived (~30 min), exchanged from the OAuth token on demand
 *
 * All Copilot API requests must use the JWT as the Bearer token plus the
 * required Copilot-Integration-Id header (handled in ai.js).
 */

import { invoke } from '@tauri-apps/api/core'

// ── Constants ─────────────────────────────────────────────────────────────────

export const COPILOT_BASE_URL = 'https://api.githubcopilot.com'
const OAUTH_STORE_KEY = '__copilot_oauth__'

// Copilot-required headers that must be added to every API request
export const COPILOT_EXTRA_HEADERS = {
  'Copilot-Integration-Id': 'vscode-chat',
  'Editor-Version': 'vscode/1.95.0',
}

// ── In-memory JWT cache ───────────────────────────────────────────────────────

/** @type {{ token: string, expiresAt: number } | null} */
let cachedJwt = null

// ── Device flow ───────────────────────────────────────────────────────────────

/**
 * Start the GitHub Device OAuth flow.
 * @returns {Promise<{ deviceCode: string, userCode: string, verificationUri: string, expiresIn: number, interval: number }>}
 */
export async function startDeviceFlow() {
  return invoke('copilot_start_device_flow')
}

/**
 * Poll GitHub for the OAuth access token after the user authorized the device.
 * Call this every `interval` seconds (from startDeviceFlow response).
 * @param {string} deviceCode
 * @returns {Promise<{ status: string, token?: string }>}
 */
export async function pollOAuthToken(deviceCode) {
  return invoke('copilot_poll_oauth_token', { deviceCode })
}

// ── Token storage ─────────────────────────────────────────────────────────────

/** Store the long-lived GitHub OAuth token securely. */
export async function storeOAuthToken(token) {
  cachedJwt = null  // invalidate cached JWT on new OAuth token
  try {
    await invoke('ai_store_key', { profileId: OAUTH_STORE_KEY, apiKey: token })
  } catch {
    // dev mode without Tauri — ignore
  }
}

/** Load the stored GitHub OAuth token, or '' if not connected. */
export async function loadOAuthToken() {
  try {
    return (await invoke('ai_load_key', { profileId: OAUTH_STORE_KEY })) ?? ''
  } catch {
    return ''
  }
}

/** Remove the stored OAuth token and clear the JWT cache. */
export async function clearOAuthToken() {
  cachedJwt = null
  try {
    await invoke('ai_delete_key', { profileId: OAUTH_STORE_KEY })
  } catch {}
}

/** Returns true if a GitHub OAuth token is stored (user is connected). */
export async function isCopilotConnected() {
  const token = await loadOAuthToken()
  return Boolean(token)
}

// ── JWT management ────────────────────────────────────────────────────────────

/**
 * Get a valid Copilot JWT, refreshing from the OAuth token if expired.
 * Uses an in-memory cache; the JWT expires in ~30 min so we refresh with a 60s buffer.
 * @throws {Error} if the user is not connected to Copilot
 */
export async function getCopilotJwt() {
  const nowSec = Math.floor(Date.now() / 1000)
  if (cachedJwt && cachedJwt.expiresAt - nowSec > 60) {
    return cachedJwt.token
  }

  const oauthToken = await loadOAuthToken()
  if (!oauthToken) {
    throw new Error('Not connected to GitHub Copilot — please log in via AI model settings.')
  }

  const result = await invoke('copilot_get_copilot_token', { oauthToken })
  cachedJwt = { token: result.token, expiresAt: result.expires_at }
  return result.token
}

/**
 * Fetch the list of models available for this Copilot subscription.
 * Automatically obtains/refreshes the JWT.
 * @returns {Promise<{ id: string, name: string }[]>}
 */
export async function fetchCopilotModels() {
  const jwt = await getCopilotJwt()
  return invoke('copilot_fetch_models', { copilotToken: jwt })
}

File: src-tauri/src/copilot.rs
use serde::{Deserialize, Serialize};

// The OAuth app client ID used by GitHub Copilot CLI / VS Code extension.
// This is a public client_id; it does not grant access without user authorization.
const GITHUB_CLIENT_ID: &str = "Iv1.b507a08c87ecfe98";

#[derive(Debug, Serialize, Deserialize)]
pub struct DeviceFlowStart {
    pub device_code: String,
    pub user_code: String,
    pub verification_uri: String,
    pub expires_in: u64,
    pub interval: u64,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct PollResult {
    /// "authorized" | "pending" | "slow_down" | "expired" | "access_denied"
    pub status: String,
    pub token: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct CopilotToken {
    pub token: String,
    /// Unix timestamp (seconds) when the JWT expires (~30 min from issue).
    pub expires_at: u64,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct CopilotModel {
    pub id: String,
    pub name: String,
}

fn http_client() -> Result<reqwest::Client, String> {
    reqwest::Client::builder()
        .user_agent("db-studio/1.0")
        .build()
        .map_err(|e| e.to_string())
}

/// Step 1 of the GitHub Device OAuth flow: obtain a user_code and verification URL.
#[tauri::command]
pub async fn copilot_start_device_flow() -> Result<DeviceFlowStart, String> {
    let client = http_client()?;
    // Use raw body so we don't need the reqwest `form` feature.
    let body = format!("client_id={}&scope=read%3Auser", GITHUB_CLIENT_ID);

    let resp = client
        .post("https://github.com/login/device/code")
        .header("Content-Type", "application/x-www-form-urlencoded")
        .header("Accept", "application/json")
        .body(body)
        .send()
        .await
        .map_err(|e| format!("Network error: {e}"))?;

    let parsed: serde_json::Value = resp
        .json()
        .await
        .map_err(|e| format!("Failed to parse GitHub response: {e}"))?;

    if let Some(err) = parsed["error"].as_str() {
        return Err(format!(
            "{}: {}",
            err,
            parsed["error_description"].as_str().unwrap_or("")
        ));
    }

    Ok(DeviceFlowStart {
        device_code: parsed["device_code"].as_str().unwrap_or("").to_string(),
        user_code: parsed["user_code"].as_str().unwrap_or("").to_string(),
        verification_uri: parsed["verification_uri"]
            .as_str()
            .unwrap_or("https://github.com/login/device")
            .to_string(),
        expires_in: parsed["expires_in"].as_u64().unwrap_or(900),
        interval: parsed["interval"].as_u64().unwrap_or(5),
    })
}

/// Step 2: poll GitHub until the user has authorized the device.
/// Call this every `interval` seconds (from the DeviceFlowStart response).
#[tauri::command]
pub async fn copilot_poll_oauth_token(device_code: String) -> Result<PollResult, String> {
    let client = http_client()?;
    let body = format!(
        "client_id={}&device_code={}&grant_type=urn%3Aietf%3Aparams%3Aoauth%3Agrant-type%3Adevice_code",
        GITHUB_CLIENT_ID, device_code
    );

    let resp = client
        .post("https://github.com/login/oauth/access_token")
        .header("Content-Type", "application/x-www-form-urlencoded")
        .header("Accept", "application/json")
        .body(body)
        .send()
        .await
        .map_err(|e| format!("Network error: {e}"))?;

    let parsed: serde_json::Value = resp
        .json()
        .await
        .map_err(|e| format!("Failed to parse response: {e}"))?;

    if let Some(token) = parsed["access_token"].as_str() {
        if !token.is_empty() {
            return Ok(PollResult {
                status: "authorized".to_string(),
                token: Some(token.to_string()),
            });
        }
    }

    let error = parsed["error"].as_str().unwrap_or("authorization_pending");
    let status = match error {
        "authorization_pending" => "pending",
        "slow_down" => "slow_down",
        "expired_token" => "expired",
        "access_denied" => "access_denied",
        _ => "pending",
    }
    .to_string();

    Ok(PollResult { status, token: None })
}

/// Exchange a GitHub OAuth token for a short-lived Copilot JWT (~30 min).
/// The JWT is required for all Copilot API calls. Refresh it before it expires.
#[tauri::command]
pub async fn copilot_get_copilot_token(oauth_token: String) -> Result<CopilotToken, String> {
    let client = http_client()?;

    let resp = client
        .get("https://api.github.com/copilot_internal/v2/token")
        .header("Authorization", format!("Bearer {}", oauth_token))
        .header("Accept", "application/json")
        .header("User-Agent", "db-studio/1.0")
        .send()
        .await
        .map_err(|e| format!("Network error: {e}"))?;

    if !resp.status().is_success() {
        let status = resp.status().as_u16();
        let body = resp.text().await.unwrap_or_default();
        return Err(format!(
            "GitHub returned {status} — OAuth token may be expired or revoked. {body}"
        ));
    }

    let parsed: serde_json::Value = resp
        .json()
        .await
        .map_err(|e| format!("Failed to parse Copilot token response: {e}"))?;

    let token = parsed["token"]
        .as_str()
        .ok_or_else(|| "No token field in Copilot response".to_string())?
        .to_string();

    let expires_at = parsed["expires_at"].as_u64().unwrap_or(0);

    Ok(CopilotToken { token, expires_at })
}

/// Fetch the list of models available to this Copilot subscription.
/// Falls back to a static list if the endpoint is unreachable.
#[tauri::command]
pub async fn copilot_fetch_models(copilot_token: String) -> Result<Vec<CopilotModel>, String> {
    let client = http_client()?;

    let resp = client
        .get("https://api.githubcopilot.com/models")
        .header("Authorization", format!("Bearer {}", copilot_token))
        .header("Content-Type", "application/json")
        .header("Copilot-Integration-Id", "vscode-chat")
        .header("Editor-Version", "vscode/1.95.0")
        .header("Accept", "application/json")
        .send()
        .await
        .map_err(|e| format!("Network error: {e}"))?;

    if !resp.status().is_success() {
        // Return the static fallback list so the UI stays functional.
        return Ok(fallback_models());
    }

    let parsed: serde_json::Value = resp
        .json()
        .await
        .map_err(|_| "Failed to parse models response".to_string())?;

    // API may return { data: [...] } or just an array at the root.
    let arr = parsed["data"]
        .as_array()
        .or_else(|| parsed.as_array())
        .cloned()
        .unwrap_or_default();

    let models: Vec<CopilotModel> = arr
        .iter()
        .filter_map(|m| {
            let id = m["id"].as_str()?.to_string();
            // Skip embedding / non-chat models
            let caps_type = m["capabilities"]["type"].as_str().unwrap_or("chat");
            if caps_type == "embeddings" {
                return None;
            }
            let name = m["name"]
                .as_str()
                .or_else(|| m["display_name"].as_str())
                .unwrap_or(&id)
                .to_string();
            Some(CopilotModel { id, name })
        })
        .collect();

    if models.is_empty() {
        Ok(fallback_models())
    } else {
        Ok(models)
    }
}

fn fallback_models() -> Vec<CopilotModel> {
    vec![
        CopilotModel { id: "gpt-4o".into(), name: "GPT-4o".into() },
        CopilotModel { id: "gpt-4o-mini".into(), name: "GPT-4o mini".into() },
        CopilotModel { id: "o3-mini".into(), name: "o3-mini".into() },
        CopilotModel { id: "claude-3.5-sonnet".into(), name: "Claude 3.5 Sonnet".into() },
        CopilotModel { id: "claude-3.7-sonnet".into(), name: "Claude 3.7 Sonnet".into() },
        CopilotModel { id: "gemini-2.0-flash-001".into(), name: "Gemini 2.0 Flash".into() },
    ]
}
